### DISCUSSION TITLE
Per file version

### DISCUSSION THREAD

**Discussion by meza** [Kira-NT/mc-publish#140] [Q&A]
Hello,

I've started using mc-publish with stonecutter projects which means that I have a bunch of jar files in my build directory targeting different loaders and minecraft versions. I'd love to specify the build directory and let mc-publish handle the uploads but what I've found is that on Modrinth for example every file belonging to the same loader gets bundled into the same "version".

While _technically_ it's correct and the logical grouping is perfect, the Modrinth user experience really suffers.

Here's a picture of how the versions look like
![image](https://github.com/user-attachments/assets/85f0cee5-cb74-4cac-b5c9-0f952ab0f60b)

When you're looking for the latest 1.21.4 for example, you click on 1.3.2 and now it suddenly has a main Download button, and 2 files attached

![image](https://github.com/user-attachments/assets/d3bd5ddc-ba34-484a-9e20-5499ac30029b)

When you click the main download button, it will download the 1.21.1 version. So now the users would have to pay very close attention to the files listed below to guess their correct version.

Long story short, my question is:

Is there a setting that would allow me to tell mc-publish to publish each file independently or would I need to do some magic in the actions?

**Comment by meza** [comment]
Seems to have the same problem on Curseforge :(

**Comment by Kira-NT** [comment]
Hi! Apologies for the delayed response; I'm currently on a break :)

This issue is less of a problem and more of a by-design behavior. `files` aren't treated as separate mods/plugins/whatever, but rather as a mod/plugin/whatever and its secondary files *(like `-dev`, `-sources`, and so on)*. So no, I wouldn't say that the result you got with your configuration is correct, because you clearly intend to upload different versions.

`mc-publish` is designed to propagate a **single** version across different platforms. And I don't really want to change that and add more switches and toggles to make it treat each file as a separate version, because doing so would drastically increase the project's complexity, while some users already struggle with the extensive configurability of this project. Moreover, if you want to publish multiple versions, you can already achieve that by simply calling `mc-publish` multiple times - that's always an option. Since GitHub Workflows are Turing complete, you can even create an improvised loop using matrices.

Given that you know in advance how many different versions you're building *(since your entire build process depends on this knowledge)*, I strongly recommend hardcoding that information into a matrix, as this would be the simplest solution. You can find an example of such a configuration for multi-loader solutions *(uploading Fabric and Forge builds of your mod alongside each other is essentially the same as uploading 2+ builds of your mod for different Minecraft versions)* [here](https://github.com/Kir-Antipov/mc-publish/discussions/29#discussioncomment-3575584).
